### PR TITLE
FEAT: implement chat and generate in RESTful client

### LINKED ISSUE #148: FEAT: implement chat and generate in RESTful client

### PR DESCRIPTION
**Pull request by jiayini1119**
Resolves #148.

### CODE REVIEW

File: xinference/client.py [hunk 1]
```diff
@@ -36,6 +36,166 @@
...
+
+class RESTfulLlamaCppModelHandle(RESTfulModelHandle):
+    def generate(
+        self, prompt: str, generate_config: Optional["LlamaCppGenerateConfig"] = None
+    ) -> Union["Completion", Iterator["CompletionChunk"]]:
+        url = f"{self._base_url}/v1/completions"
+        if generate_config is None:
+            request_body = {"model": self._model_uid, "prompt": prompt}
+        else:
+            dict = {}
```
Comment (pangyoki) [inline xinference/client.py:102]: Not important. Better not to use python's built-in type name `dict`.

File: xinference/client.py [hunk 2]
```diff
@@ -36,6 +36,166 @@
...
+        else:
+            if system_prompt is not None:
+                chat_history.insert(0, {"role": "system", "content": system_prompt})
+
+        chat_history.append({"role": "user", "content": prompt})
+
+        if generate_config is None:
+            request_body = {"model": self._model_uid, "messages": chat_history}
+        else:
+            dict = {}
```
Comment (pangyoki) [inline xinference/client.py:146]: Not important. Better not to use python's built-in type name `dict`.

File: xinference/client.py [hunk 3]
```diff
@@ -36,6 +36,166 @@
...
+
+        if chat_history is None:
+            chat_history = []
+
+        chat_history.append({"role": "user", "content": prompt})
+
+        if generate_config is None:
+            request_body = {"model": self._model_uid, "messages": chat_history}
+        else:
+            dict = {}
```
Comment (pangyoki) [inline xinference/client.py:181]: Not important. Better not to use python's built-in type name `dict`.

File: xinference/client.py [hunk 4]
```diff
@@ -36,6 +36,166 @@
...
+        response_data = response.json()
+        return response_data
+
+
+class RESTfulChatglmCppChatModelHandle(RESTfulModelHandle):
+    def chat(
+        self,
+        prompt: str,
+        chat_history: Optional[List["ChatCompletionMessage"]] = None,
+        generate_config: Optional["LlamaCppGenerateConfig"] = None,
```
Comment (pangyoki) [inline xinference/client.py:169]: `LlamaCppGenerateConfig` -> `ChatglmCppGenerateConfig`

**Review by pangyoki — APPROVED** [review]
LGTM

File: xinference/core/restful_api.py [hunk 5]
```diff
@@ -32,7 +32,7 @@
 logger = logging.getLogger(__name__)
 
 max_tokens_field = Field(
-    default=16, ge=1, le=2048, description="The maximum number of tokens to generate."
+    default=512, ge=1, le=2048, description="The maximum number of tokens to generate."
```
Comment (UranusSeven) [inline xinference/core/restful_api.py:35]:
```suggestion
    default=128, ge=1, le=2048, description="The maximum number of tokens to generate."
```

File: xinference/tests/test_client.py [hunk 6]
```diff
@@ -35,3 +35,34 @@ async def test_sync_client(setup):
 
     client.terminate_model(model_uid=model_uid)
     assert len(client.list_models()) == 0
+
+
+@pytest.mark.asyncio
+async def test_RESTful_client(setup):
+    endpoint, _ = setup
```
Comment (UranusSeven) [inline xinference/tests/test_client.py:42]: Could u also test `describe_model`?

**Pull request merged**